### PR TITLE
 Fix TreeList getSelectedRowKeys behaviour in leavesOnly case

### PR DESCRIPTION
**Pull request by sometimesElen**

### CODE REVIEW

**Comment by sometimesElen** [comment]
Require @lukyanovas to review the deprecations in doccomments

File: js/ui/tree_list/ui.tree_list.data_source_adapter.js [hunk 1]
```diff
@@ -560,7 +560,7 @@ DataSourceAdapter = DataSourceAdapter.inherit((function() {
 
                 node && treeListCore.foreachNodes([node], function(childNode) {
                     if(!childNode.children.length && (!callBack || callBack(childNode))) {
-                        result.push(childNode.key);
+                        result.indexOf(childNode.key) < 0 && result.push(childNode.key);
```
Comment (vconst) [inline js/ui/tree_list/ui.tree_list.data_source_adapter.js]: The indexOf is very hard op[eration for large amount of nodes

**Review by vconst — CHANGES_REQUESTED** [review]

**Review by vconst — APPROVED** [review]

File: js/ui/tree_list/ui.tree_list.selection.js [hunk 2]
```diff
@@ -416,27 +430,39 @@ treeListCore.registerModule("selection", extend(true, {}, selectionModule, {
...
-                            });
+                    mode = mode || "excludeRecursive";
+
+                    if(dataController) {
+                        if(mode === true) {
+                            errors.log("W0002", "dxTreeList", "getSelectedRowKeys(leavesOnly)", "18.1", "Use the 'getSelectedRowKeys(mode)' method with a string parameter instead");
                         }
 
-                        if(mode === "all") {
+                        if(this.isRecursiveSelection() && mode !== "excludeRecursive") {
```
Comment (Alyar666) [inline js/ui/tree_list/ui.tree_list.selection.js:459]: I think it's better to check on 'includeRecursive': `mode === "includeRecursive"`

**Review by Alyar666 — CHANGES_REQUESTED** [review]

File: js/ui/tree_list/ui.tree_list.selection.js (hunk 2 above)
Comment (sometimesElen) [reply js/ui/tree_list/ui.tree_list.selection.js:459]: After refix this string can look like this or ```mode === "includeRecursive" || mode === "leavesOnly"```

**Review by Alyar666 — APPROVED** [review]

**Pull request merged**